network/p2p/dns/cache.go

```diff
@@ -91,3 +97,29 @@ func (c *cache) updateTXTCache(txt string, addr []string) {
 		timestamp: runtimeNano(),
 	}
 }
+
+// invalidateIPCacheEntry atomically invalidates ip cache entry. Boolean variable determines whether invalidation
+// is successful.
+func (c *cache) invalidateIPCacheEntry(domain string) bool {
+	c.Lock()
```

Comment:
Nit: but how about this?
```suggestion
	c.RLock()
	_, exists := c.ipCache[domain]
	c.RUnlock()
	
	if !exists {
	    return false
	}
	delete(c.ipCache, domain) // delete is idempotent and once in a while it is ok to report as exist even when it may have been deleted avoid the cost of a write lock
```

Reply:
@vishalchangrani @huitseeker I applied this comment and refactored similar places, however, now the code is vulnerable to concurrent map access which will result in crash failure. Currently, the DNS resolver interacting with cache follows a causality in operations, e.g., it does not perform invalidation concurrently with resolving. However, if this behavior is changed, later on, this race condition will emerge as a crash failure. Just wanted to highlight it.